src/Compilers/Core/Portable/Syntax/GreenNode.NodeEnumerable.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis;
+
+internal abstract partial class GreenNode
+{
+    [NonCopyable]
+    public ref struct NodeEnumerable(GreenNode node)
+    {
+        public readonly Enumerator GetEnumerator()
+            => new Enumerator(node);
+
+        [NonCopyable]
+        public ref struct Enumerator
+        {
+            private readonly ArrayBuilder<Syntax.InternalSyntax.ChildSyntaxList.Enumerator> _stack;
+
+            private bool _started;
+            private GreenNode _current;
+
+            public Enumerator(GreenNode node)
+            {
+                _current = node;
+                _stack = ArrayBuilder<Syntax.InternalSyntax.ChildSyntaxList.Enumerator>.GetInstance();
+                _stack.Push(node.ChildNodesAndTokens().GetEnumerator());
+            }
+
+            public readonly void Dispose()
+                => _stack.Free();
```

Comment:
We're doing custom iteration of enumerators (from `GetEnumerator()`) in this type. Enumerators should normally get disposed. How do we know it's safe not to dispose them?

Reply:
I'm not sure i understand the question.  It's safe to not dispose this enumerator.  You'll just end up with some garbage in that case.  The same as with the prior IEnumerator.

Reply:
resolved offline.  the enumerators we push onto the stack are not tehmselves disposable.  This is a common pattern for our struct enumerators (both internal and public).